File: app/src/main/java/com/simplegames/chris/rockpaperscissors20/SinglePlayer.java
package com.simplegames.chris.rockpaperscissors20;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.AnimationDrawable;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.DecelerateInterpolator;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Random;

import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.core.widget.ImageViewCompat;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;

import com.google.android.material.card.MaterialCardView;

public class SinglePlayer extends AppCompatActivity {

    LinearLayout Arrow;
    MaterialCardView scissorsButton, rockButton, paperButton, buttonMenu, resultHolder;
    ConstraintLayout background;
    FrameLayout aiResult;
    ImageView aiChoice;
    TextView resultText, lossesScore, drawsScore, winsScore;
    AnimationDrawable aiAnimation;
    int resetCounter = 0;
    boolean isPlaying = false;

    String aiReturn = "";
    String playerReturn = "";
    String winnerReturn = "";

    int buttonHeight;
    int resultHeight;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Values.darkTheme){
            setTheme(R.style.DarkTheme);
        }else {
            setTheme(R.style.LightTheme);
        }
        setContentView(R.layout.activity_single_player);
        if (this.getTheme().equals(R.style.LightTheme)){
            aiChoice = findViewById(R.id.aiChoice);
            ImageViewCompat.setImageTintList(aiChoice, ColorStateList.valueOf(R.attr.GroupColour));
        }

        aiResult = findViewById(R.id.aiResult);

        rockButton = findViewById(R.id.rockButton);
        paperButton = findViewById(R.id.paperButton);
        scissorsButton = findViewById(R.id.scissorsButton);

        background = findViewById(R.id.background);
        buttonMenu = findViewById(R.id.buttonMenu);
        resultHolder = findViewById(R.id.resultHolder);
        aiChoice = findViewById(R.id.aiChoice);
        aiAnimation = (AnimationDrawable) aiChoice.getDrawable();
        aiAnimation.setEnterFadeDuration(0);
        aiAnimation.setExitFadeDuration(0);

        final Handler animPlay = new Handler();
        resultText = findViewById(R.id.resultText);
        lossesScore = findViewById(R.id.lossesScore);
        drawsScore = findViewById(R.id.drawsScore);
        winsScore = findViewById(R.id.winsScore);
        openGame();
        sendBackgroundRequest();
        resetScoreCounter();
        singlePlayerIN();

        /**
         *
         *          What happens when buttons are pressed.
         *
         */
        buttonMenu.setOnClickListener(v -> {
            if (!isPlaying){
                buttonMenu.setClickable(false);
                Vibrations.openGameMode(SinglePlayer.this);
                UIElements.animate(aiResult, "translationY", 60, 100, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(aiResult, "alpha", 0, 100, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(resultHolder, "translationY", -resultHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(buttonMenu, "alpha", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(rockButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(paperButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(scissorsButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                Handler handler = new Handler();
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        Goto.settingsNew(SinglePlayer.this);
                        finish();
                        SinglePlayer.this.overridePendingTransition(0,0);
                    }
                }, Values.animationSpeed);

            }else {
                Vibrations.UnavailableVibration(getApplicationContext());
                LayoutInflater layoutInflater = getLayoutInflater();
                View cantExitView = layoutInflater.inflate(R.layout.toast_in_game, (ViewGroup)findViewById(R.id.cantExitToast));
                Toast cantExitToast = Toast.makeText(SinglePlayer.this, "Can't Leave", Toast.LENGTH_SHORT);
                cantExitToast.setGravity(Gravity.CENTER,0,0);
                cantExitToast.setView(cantExitView);
                cantExitToast.show();
            }
        });

        rockButton.setOnClickListener(v -> {
            buttonHeight = rockButton.getHeight();
            if (!isPlaying){
                isPlaying = true;
                Vibrations.SPButtonSelectedVibration(getApplicationContext());
                playerReturn = "Rock";
                playAnimation();
                aiPick();
                winnerChosen();
                resultAnimation();
                //UIElements.animate(rockButton, "translationY", -50, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(paperButton, "translationY", buttonHeight, 50, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(scissorsButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                animPlay.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        aiAnimation.stop();
                        aiShow();
                        isPlaying = false;
                        //UIElements.animate(rockButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                        UIElements.animate(paperButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                        UIElements.animate(scissorsButton, "translationY", 0, 50, Values.animationSpeed, new DecelerateInterpolator(3));
                    }
                },6050);
            } else {
                Vibrations.SPAlreadyPlayingVibration(getApplicationContext());
            }


        });
        paperButton.setOnClickListener(v -> {
            buttonHeight = rockButton.getHeight();
            if (!isPlaying){
                isPlaying = true;
                Vibrations.SPButtonSelectedVibration(getApplicationContext());
                playerReturn = "Paper";
                playAnimation();
                aiPick();
                winnerChosen();
                resultAnimation();
                UIElements.animate(rockButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(scissorsButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                animPlay.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        aiAnimation.stop();
                        aiShow();
                        isPlaying = false;
                        UIElements.animate(rockButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                        UIElements.animate(scissorsButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                    }
                },6050);
            } else {
                Vibrations.SPAlreadyPlayingVibration(getApplicationContext());
            }

        });
        scissorsButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                buttonHeight = rockButton.getHeight();
                if (!isPlaying){
                    isPlaying = true;
                    Vibrations.SPButtonSelectedVibration(getApplicationContext());
                    playerReturn = "Scissors";
                    playAnimation();
                    aiPick();
                    winnerChosen();
                    resultAnimation();
                    UIElements.animate(rockButton, "translationY", buttonHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                    UIElements.animate(paperButton, "translationY", buttonHeight, 50, Values.animationSpeed, new DecelerateInterpolator(3));
                    animPlay.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            aiAnimation.stop();
                            aiShow();
                            isPlaying = false;
                            UIElements.animate(rockButton, "translationY", 0, 50, Values.animationSpeed, new DecelerateInterpolator(3));
                            UIElements.animate(paperButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                        }
                    },6050);
                } else {
                    Vibrations.SPAlreadyPlayingVibration(getApplicationContext());
                }

            }
        });


    }

    @Override
    protected void onStart() {
        super.onStart();
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                Values.currentActivity = "SinglePlayer";
            }
        }, 2000);
    }

    public void playAnimation(){
        aiChoice.setImageDrawable(getResources().getDrawable(R.drawable.ai_picking));
        AnimationDrawable pickingAnimation = (AnimationDrawable)aiChoice.getDrawable();
        pickingAnimation.start();
    }
    public void aiPick(){
        Random r = new Random();
        int choice = r.nextInt(100)+1;
        if (choice < Values.aiRockPercent){
            aiReturn = "Rock";
        }else if (choice < Values.aiRockPercent + Values.aiPaperPercent){
            aiReturn = "Paper";
        }else if (choice < Values.aiRockPercent + Values.aiPaperPercent + Values.aiScissorsPercent){
            aiReturn = "Scissors";
        }else{aiPick();}
    }
    public void aiShow(){
        if (aiReturn == "Rock"){
            aiChoice.setImageResource(R.drawable.icon_rock);
        } else if (aiReturn == "Paper"){
            aiChoice.setImageResource(R.drawable.icon_paper);
        } else if (aiReturn == "Scissors"){
            aiChoice.setImageResource(R.drawable.icon_scissors);
        }
    }
    public void winnerChosen(){
        if (aiReturn == "Rock" && playerReturn == "Rock"){
            winnerReturn = "Draw";
        } else if (aiReturn == "Rock" && playerReturn == "Paper"){
            winnerReturn = "PlayerWin";
        } else if (aiReturn == "Rock" && playerReturn == "Scissors"){
            winnerReturn = "AiWin";
        }

        if (aiReturn == "Paper" && playerReturn == "Rock"){
            winnerReturn = "AiWin";
        } else if (aiReturn == "Paper" && playerReturn == "Paper"){
            winnerReturn = "Draw";
        } else if (aiReturn == "Paper" && playerReturn == "Scissors"){
            winnerReturn = "PlayerWin";
        }

        if (aiReturn == "Scissors" && playerReturn == "Rock"){
            winnerReturn = "PlayerWin";
        } else if (aiReturn == "Scissors" && playerReturn == "Paper"){
            winnerReturn = "AiWin";
        } else if (aiReturn == "Scissors" && playerReturn == "Scissors"){
            winnerReturn = "Draw";
        }
    }
    public void winnerDisplay(){
        if (winnerReturn == "Draw"){
            resultText.setText("Draw");
            Values.draws++;
            drawsScore.setText("Draws: "+ Values.draws);
            //aiSmarter();
        } else if (winnerReturn == "PlayerWin"){
            resultText.setText("You Win");
            Values.wins++;
            winsScore.setText("Wins: "+ Values.wins);
            //aiSmarter();
        } else if (winnerReturn == "AiWin"){
            resultText.setText("You Lose");
            Values.losses++;
            lossesScore.setText("Losses: "+ Values.losses);
            //aiSmarter();
        }
        if (Values.losses < 0 || Values.draws < 0 || Values.wins < 0){
        }
    }

    /**
     *          Plays the animation of the result box
     *          and displays the correct winner.
     *
     */
    public void resultAnimation(){
        Handler outDelay = new Handler();
        final Handler changeText = new Handler();
        final Handler inDelay = new Handler();
        //final Animation out = AnimationUtils.loadAnimation(this, R.anim.result_animation_out);
        //final Animation in = AnimationUtils.loadAnimation(this, R.anim.result_animation_in);
        outDelay.postDelayed(new Runnable() {
            @Override
            public void run() {
                UIElements.animate(resultHolder, "translationY", -resultHeight, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                changeText.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        inDelay.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                winnerDisplay();
                                UIElements.animate(resultHolder, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                            }
                        }, 3634);
                    }
                }, 400);
            }
        }, 2016);


    }

    public void openGame(){
        try{
            lossesScore.setText("Losses: "+ Values.losses);
            drawsScore.setText("Draws: "+ Values.draws);
            winsScore.setText("Wins: "+ Values.wins);
        }catch (Exception loadFail){
            Toast.makeText(this, "Failed to Load save", Toast.LENGTH_SHORT).show();
        }
    }
    public void sendBackgroundRequest(){
        ConstraintLayout background = findViewById(R.id.background);
        switch (Values.SPBackgroundNumber){
            case 0:
                UIElements.twoPartGradient(background,null, ContextCompat.getColor(this, R.color.wintersDayTL),
                        ContextCompat.getColor(this, R.color.wintersDayBR),0f);
                return;
            case 1:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.shallowLakeTL),
                        ContextCompat.getColor(this, R.color.shallowLakeBR), 0f);
                return;
            case 2:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.tropicalOceanTL),
                        ContextCompat.getColor(this, R.color.tropicalOceanBR), 0f);
                return;
            case 3:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.greenGrassTL),
                        ContextCompat.getColor(this, R.color.greenGrassBR), 0f);
                return;
            case 4:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.sunshineTL),
                        ContextCompat.getColor(this, R.color.sunshineBR), 0f);
                return;
            case 5:
                UIElements.twoPartGradient(background, null,ContextCompat.getColor(this, R.color.forestTL),
                        ContextCompat.getColor(this, R.color.forestBR), 0f);
                return;
            case 6:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.atmosphereTL),
                        ContextCompat.getColor(this, R.color.atmosphereBR), 0f);
                return;
            case 7:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.purpleHazeTL),
                        ContextCompat.getColor(this, R.color.purpleHazeBR), 0f);
                return;
            case 8:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.coldNightTL),
                        ContextCompat.getColor(this, R.color.coldNightBR), 0f);
                return;
            case 9:
                UIElements.twoPartGradient(background,null,ContextCompat.getColor(this, R.color.eternalSpaceTL),
                        ContextCompat.getColor(this, R.color.eternalSpaceBR), 0f);
        }
    }
    public void resetScoreCounter(){
        if (Values.resetScoreSP){
            Values.wins = resetCounter;
            Values.draws = resetCounter;
            Values.losses = resetCounter;
            lossesScore.setText("Losses: " + Values.losses);
            drawsScore.setText("Draws: " + Values.draws);
            winsScore.setText("Wins: " + Values.wins);
            Values.resetScoreSP = false;
        }
    }

    public void singlePlayerIN(){
        rockButton.post(new Runnable() {
            @Override
            public void run() {
                buttonHeight = rockButton.getHeight();
                resultHeight = resultHolder.getHeight() + resultHolder.getPaddingTop();
                Log.e("NOTE", ""+buttonHeight);
                aiResult.setTranslationY(60);
                aiResult.setAlpha(0f);
                resultHolder.setTranslationY(-resultHeight);
                buttonMenu.setAlpha(0f);
                rockButton.setTranslationY(buttonHeight);
                paperButton.setTranslationY(buttonHeight);
                scissorsButton.setTranslationY(buttonHeight);
                UIElements.animate(aiResult, "translationY", 0, 100, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(aiResult, "alpha", 1, 100, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(buttonMenu, "alpha", 1, 100, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(resultHolder, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(rockButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(paperButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                UIElements.animate(scissorsButton, "translationY", 0, 0, Values.animationSpeed, new DecelerateInterpolator(3));
                Values.currentActivity = "SinglePlayer";
            }
        });

    }

    /*public void aiSmarter(){
        if (playerReturn == "Rock"){
            aiRockPercent = aiRockPercent - 11;
            aiPaperPercent = aiPaperPercent + 22;
            aiScissorsPercent = aiScissorsPercent - 11;
            aiSmarterValueCheck();
        }else if (playerReturn == "Paper"){
            aiRockPercent = aiRockPercent - 11;
            aiPaperPercent = aiPaperPercent - 11;
            aiScissorsPercent = aiScissorsPercent + 22;
            aiSmarterValueCheck();
        }else if (playerReturn == "Scissors"){
            aiRockPercent = aiRockPercent + 22;
            aiPaperPercent = aiPaperPercent - 11;
            aiScissorsPercent = aiScissorsPercent - 11;
            aiSmarterValueCheck();
        }else{
            Toast.makeText(this,"'AI Smarter' Failed",Toast.LENGTH_SHORT).show();
            aiSmarterValueCheck();
        }
        if (winnerReturn == "PlayerWin" || winnerReturn == "Draw" || winnerReturn == "AIWin"){

        }
    }

    public void aiSmarterValueCheck(){
        if (aiRockPercent > 100){
            aiRockPercent = 100;
            aiPaperPercent = 0;
            aiScissorsPercent = 0;
        } else if (aiPaperPercent > 100){
            aiRockPercent = 0;
            aiPaperPercent = 100;
            aiScissorsPercent = 0;
        } else if (aiScissorsPercent > 100){
            aiRockPercent = 0;
            aiPaperPercent = 0;
            aiScissorsPercent = 100;
        }else if (aiRockPercent < 0){
            aiRockPercent = 0;
        }else if (aiPaperPercent < 0){
            aiPaperPercent = 0;
        }else if (aiScissorsPercent < 0){
            aiScissorsPercent = 0;
        }
        else{
            //aiRockPercent = 33;
            //aiPaperPercent = 33;
            //aiScissorsPercent = 33;
        }
    }*/


    @Override
    public void onBackPressed(){
    }

}

File: app/src/main/java/com/simplegames/chris/rockpaperscissors20/SinglePlayer.kt
package com.simplegames.chris.rockpaperscissors20

import android.content.Intent
import android.graphics.drawable.AnimationDrawable
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.View
import android.view.animation.DecelerateInterpolator
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.res.ResourcesCompat
import com.google.android.material.card.MaterialCardView
import java.util.Random
import java.util.Timer
import kotlin.concurrent.timerTask

class SinglePlayer : AppCompatActivity() {

    //Create Screen Values
    private lateinit var buttonRock: MaterialCardView
    private lateinit var buttonPaper: MaterialCardView
    private lateinit var buttonScissors: MaterialCardView
    private lateinit var buttonMenu: MaterialCardView

    private lateinit var background: ImageView
    private lateinit var opponentChoiceImage: ImageView
    private lateinit var opponentChoiceHolder: ConstraintLayout
    private lateinit var roundResult: TextView
    private lateinit var roundResultHolder: MaterialCardView

    private lateinit var opponentAnimation: AnimationDrawable
    private var isPlaying: Boolean = false
    private var opponentChoice: String = ""
    private var playerChoice: String = ""


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //Toast.makeText(this, "Started", Toast.LENGTH_SHORT).show()
        //Set Screen Theme
        if (Values.darkThemeEnabled) {
            setTheme(R.style.DarkTheme)
        } else {
            setTheme(R.style.LightTheme)
        }

        //Set Content View
        setContentView(R.layout.activity_single_player)


        //Set Screen Values
        buttonMenu = findViewById(R.id.buttonMenu)
        buttonRock = findViewById(R.id.rockButton)
        buttonPaper = findViewById(R.id.paperButton)
        buttonScissors = findViewById(R.id.scissorsButton)

        background = findViewById(R.id.background)
        opponentChoiceImage = findViewById(R.id.aiChoice)
        opponentChoiceHolder = findViewById(R.id.aiResult)
        roundResult = findViewById(R.id.resultText)
        roundResultHolder = findViewById(R.id.resultHolder)

        opponentAnimation = opponentChoiceImage.drawable as AnimationDrawable
        UIElements.setBackground(this, background, UIElements.getBackgroundColours(this), 0f, 0)

        //Animate Elements In
        animateIn()
        Log.e("ERR RPS: ", "Created")
        //Button Presses
        buttonMenu.setOnClickListener {
            if (isPlaying) {
                vibrate(this, VibrationType.ERROR)
            } else {
                //Stop multiple clicks
                buttonMenu.isClickable = false

                //Vibrate
                vibrate(this, VibrationType.MEDIUM)

                //Animate UI Out
                UIElements.animate(
                    buttonMenu,
                    "alpha",
                    0,
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    buttonRock,
                    "translationY",
                    (buttonRock.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    buttonPaper,
                    "translationY",
                    (buttonPaper.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    buttonScissors,
                    "translationY",
                    (buttonScissors.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    roundResultHolder,
                    "translationY",
                    -(roundResultHolder.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    opponentChoiceHolder,
                    "translationY",
                    60,
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    opponentChoiceHolder,
                    "alpha",
                    0,
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )

                //Open Settings
                Timer().schedule(timerTask {
                    val pageSettings = Intent(this@SinglePlayer, Settings::class.java)
                    startActivity(pageSettings.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION))
                    finish()
                }, Values.animationSpeed.toLong())
            }
        }

        buttonRock.setOnClickListener {
            if (isPlaying) {
                vibrate(this, VibrationType.ERROR)
            } else {
                //Set isPlaying
                isPlaying = true
                playerChoice = "Rock"

                //Vibrate
                vibrate(this, VibrationType.WEAK)

                //Hide Other Options
                UIElements.animate(
                    buttonPaper,
                    "translationY",
                    (buttonPaper.height + UIElements.dpToFloat(10f)),
                    50,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    buttonScissors,
                    "translationY",
                    (buttonScissors.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )

                //Get Results
                opponentPick()
            }
        }

        buttonPaper.setOnClickListener {
            if (isPlaying) {
                vibrate(this, VibrationType.ERROR)
            } else {
                //Set isPlaying
                isPlaying = true
                playerChoice = "Paper"

                //Vibrate
                vibrate(this, VibrationType.WEAK)

                //Hide Other Options
                UIElements.animate(
                    buttonRock,
                    "translationY",
                    (buttonPaper.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    buttonScissors,
                    "translationY",
                    (buttonScissors.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )

                //Get Results
                opponentPick()
            }
        }

        buttonScissors.setOnClickListener {
            if (isPlaying) {
                vibrate(this, VibrationType.ERROR)
            } else {
                //Set isPlaying
                isPlaying = true
                playerChoice = "Scissors"

                //Vibrate
                vibrate(this, VibrationType.WEAK)

                //Hide Other Options
                UIElements.animate(
                    buttonRock,
                    "translationY",
                    (buttonPaper.height + UIElements.dpToFloat(10f)),
                    0,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )
                UIElements.animate(
                    buttonPaper,
                    "translationY",
                    (buttonScissors.height + UIElements.dpToFloat(10f)),
                    50,
                    Values.animationSpeed,
                    DecelerateInterpolator(3f)
                )

                //Get Results
                opponentPick()
            }
        }
    }

    private fun playOpponentAnimation() {
        //Play Picking Animation
        opponentChoiceImage.setImageDrawable(ResourcesCompat.getDrawable(resources, R.drawable.ai_picking, null))
        val pickingAnimation = opponentChoiceImage.drawable as AnimationDrawable
        pickingAnimation.start()

        //Hide Result Holder
        Handler().postDelayed({
            UIElements.animate(
                roundResultHolder,
                "translationY",
                -(roundResultHolder.height + UIElements.dpToFloat(10f)),
                0,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
        }, 2000)

        //Display Winner
        Handler().postDelayed({
            displayWinner()
            pickingAnimation.stop()
            isPlaying = false
        }, 5900)
    }

    private fun opponentPick() {
        val random = Random()
        val choice = random.nextInt(3)

        opponentChoice = when (choice) {
            0 -> {
                "Rock"
            }

            1 -> {
                "Paper"
            }

            else -> {
                "Scissors"
            }
        }

        //Show Animation
        playOpponentAnimation()
    }

    private fun displayOpponentChoice() {
        when (opponentChoice) {
            "Rock" -> {
                opponentChoiceImage.setImageResource(R.drawable.icon_rock)
            }

            "Paper" -> {
                opponentChoiceImage.setImageResource(R.drawable.icon_paper)
            }

            "Scissors" -> {
                opponentChoiceImage.setImageResource(R.drawable.icon_scissors)
            }

            else -> {
                opponentChoiceImage.setImageResource(R.drawable.icon_app)
            }
        }
    }

    private fun displayWinner() {
        displayOpponentChoice()
        if (opponentChoice == playerChoice) {
            roundResult.text = "Draw"
            Values.draws++
        } else if ((opponentChoice == "Rock" && playerChoice == "Scissors")
            || (opponentChoice == "Paper" && playerChoice == "Rock")
            || (opponentChoice == "Scissors" && playerChoice == "Paper")
        ) {
            roundResult.text = "You Lose"
            Values.losses++
        } else {
            roundResult.text = "You Win"
            Values.wins++
        }

        //Animate Back In
        UIElements.animate(
            buttonRock,
            "translationY",
            0,
            50,
            Values.animationSpeed,
            DecelerateInterpolator(3f)
        )
        UIElements.animate(
            buttonPaper,
            "translationY",
            0,
            0,
            Values.animationSpeed,
            DecelerateInterpolator(3f)
        )
        UIElements.animate(
            buttonScissors,
            "translationY",
            0,
            50,
            Values.animationSpeed,
            DecelerateInterpolator(3f)
        )
        UIElements.animate(
            roundResultHolder,
            "translationY",
            0,
            0,
            Values.animationSpeed,
            DecelerateInterpolator(3f)
        )
    }

    private fun animateIn() {
        Log.e("ERR RPS: ", "Animating In")
        background.post {
            Log.e("ERR RPS: ", "UI There")
            //Set Elements Initial Positions/Alphas
            buttonMenu.alpha = 0f
            buttonRock.translationY = ((buttonRock.height + UIElements.dpToFloat(10f)).toFloat())
            buttonPaper.translationY = ((buttonPaper.height + UIElements.dpToFloat(10f)).toFloat())
            buttonScissors.translationY =
                ((buttonScissors.height + UIElements.dpToFloat(10f)).toFloat())
            opponentChoiceHolder.translationY = UIElements.dpToFloat(60f).toFloat()
            opponentChoiceHolder.alpha = 0f
            roundResultHolder.translationY =
                (-(roundResultHolder.height + UIElements.dpToFloat(10f)).toFloat())

            //Set Elements VISIBLE
            buttonMenu.visibility = View.VISIBLE
            buttonRock.visibility = View.VISIBLE
            buttonPaper.visibility = View.VISIBLE
            buttonScissors.visibility = View.VISIBLE
            opponentChoiceHolder.visibility = View.VISIBLE
            roundResultHolder.visibility = View.VISIBLE

            //Animate Elements
            UIElements.animate(
                buttonMenu,
                "alpha",
                1,
                0,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
            UIElements.animate(
                buttonRock,
                "translationY",
                0,
                0,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
            UIElements.animate(
                buttonPaper,
                "translationY",
                0,
                0,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
            UIElements.animate(
                buttonScissors,
                "translationY",
                0,
                0,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
            UIElements.animate(
                opponentChoiceHolder,
                "translationY",
                0,
                100,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
            UIElements.animate(
                opponentChoiceHolder,
                "alpha",
                1,
                100,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )
            UIElements.animate(
                roundResultHolder,
                "translationY",
                0,
                0,
                Values.animationSpeed,
                DecelerateInterpolator(3f)
            )

            //Set currentActivity
            Values.currentActivity = "SinglePlayer"
        }
    }

}